### ISSUE TITLE
questions about using code base: Monet-pytorch

### ISSUE THREAD

**Issue by ling-k** [baudm/MONet-pytorch#11]
Hi Darwin, I hope you are doing well. I have been using your codebase Genesis for several days but I got into some problems when I use my own dataset. In particular, below is the screenshot of my result when I use Monet. The desired behavior is that each object is mapped to one corresponding mask and representation slot, which is indeed the case when I use the dSprite dataset. However, when I use my own dataset, all objects are mapped to the same mask and the same slot. Could you please point out what's going on and which parameters I should change? Thanks in advance. 
![image](https://user-images.githubusercontent.com/38346659/118526735-8b559900-b70e-11eb-83e6-1d7f173e4ef3.png)

**Comment by baudm** [comment]
Hi @ling-k. Just a disclaimer that I'm not an author of the MONet paper. However, in implementing it, I had formed some intuitions on why it works on the dSprites and CLEVR datasets: the loss function of the decoder is minimized by maximizing **m<sub>k</sub>** and by minimizing the difference between **x** (input image) and μ<sub>θ</sub> (decoder output) per pixel.

Overall, this loss function is minimized by predicting **m<sub>k</sub>** = 1 **only** for pixels with similar or very close colors. This is my intuition about how the clustering works. In other words, the model learns to cluster the data by color, and it will fail for data wherein the objects have similar or close RGB values. I'm not sure if it would work, but you can try a color space wherein the separation between the raw triplet values (e.g. RGB, HSV) of any two images are farther apart.

If you can still modify your dataset, try using colors which are farther apart in the RGB color space.

Hope this helps.

**Comment by ling-k** [comment]
Thank you so much for your advice. I will try to change my dataset.
